### ISSUE TITLE
Quantitative testing: Round 1 of testing and squashing natural language false positives

### ISSUE THREAD

**Issue by RedXanadu** [coreruleset/coreruleset#3392]
This is a continuation of the [quantitative testing project from the 2023 Dev Retreat](https://github.com/coreruleset/coreruleset/wiki/DevRetreat23ProjectQuantitativeTesting).

### Goals

- We want to make it **faster** and **easier** to tune a fresh CRS installation in front of an arbitrary web application.
- To achieve this, we want to experience **fewer false positives** with natural language user input.
- We agreed that we want the _worst_ offending rules to be fixed for the v4.0.0 release.

### Proposed solution

We have a shell script that automatically tests a series of natural language sentences (English language snippets (news headlines) taken from a University of Leipzig project, [here](https://wortschatz.uni-leipzig.de/en/download)) against the CRS testing container setup (the docker-compose configuration in the CRS testing directory).

### Initial results

GitHub has limited support for tables (it isn't spreadsheet software, after all), so below is a screenshot of the initial testing results.

Tested using the latest CRS v4.0/dev branch.

Tested using the full 10,000 sample payloads (sentences) from the corpus _eng_news_2020_10K_.

**Blue highlighting** = New rule in v4.0 (so rule comparison to v3.3.5 is less meaningful)

_Test date and time: 29 Nov, 00:51 UTC_

![image](https://github.com/coreruleset/coreruleset/assets/6837454/664b629f-4d74-412b-bd0f-f27147326c0e)

### Open sub-issues

The following sub-issues are open in relation to this main task:

- #3396
- #3397
- #3511
- #3526
- #3506

### Next steps: How to make this ready for CRSv4:

The following fixes are seen as essential. It's PL2 with a frequency of above 100 and PL1 with a frequency of above 10 in 10000 payloads.

- [x] Open an issue to investigate FPs with rule 942110 (PL2)
- [x] Open an issue to investigate FPs with rule 932240 (PL2)
- [x] Open an issue to investigate FPs with rule 932236 (PL2)
- [x] Open an issue to investigate FPs with rule 920240 (PL1) (Update: add to this issue: "next step: we agree move to PL 2 and rethink in the future", and open a PR to move the rule to PL2 if there's time) (This has already been done, now)
- [x] Open an issue to investigate FPs with rule 932380 (PL1)

**Comment by dune73** [comment]
This is awesome.

When I sum up the columns, I get 165 FPs for CRSv4-RC2 at PL1 and 0 for CRS 3.3.5 at PL1.

We can now ask the questions just how much is acceptable. I think we need to define one or two thresholds per paranoia level.
* The number of FPs for an individual rule
* optional: The number of FPs for the entire rule set

We have 31 rules above. 9 of them are PL1. 

Hypothetical thresholds for individual rule: **PL1 of 0 and PL2 of 10** would mean we need to solve 9 FPs for PL1 and 10 FPs for PL2 (this includes 942100 at 4 FPs where we can not really do anything).

It seems CRS 3.3.5 was OK with 942130 PL2 at 83 FPs per 10K requests. So we could say we are more lenient and propose the following thresholds for individual rules: **PL1 of 1 and PL2 of 50**. That would mean 8 PL1 rules (including 942100) and 5 PL2 rules.

I do not think we should release with a PL2 threshold higher than 50 for a PL2 rule. And at least the lenient approach with these numbers sounds doable.

**Comment by RedXanadu** [comment]
@dune73 Thanks!

Using the original numbers (more difficult to do relying only on the diff. values in the screenshot above), the totals I get are:

- v4.0.0-rc2
  - PL 1: 165
  - PL 2: 2849
- v3.3.5
  - PL 1: 141
  - PL 2: 1852

I'm including the original numbers for completeness.

```
-= v3.3.5 =-
Freq.   ID #   Paranoia Level
   1490 942110 PL2
    130 920240 PL1
     90 942400 PL2
     83 942130 PL2
     68 942200 PL2
     25 942440 PL2
     22 942390 PL2
     19 932200 PL2
     14 942430 PL2
     12 942260 PL2
      8 942410 PL2
      7 942150 PL2
      5 942380 PL2
      4 942180 PL2
      4 942100 PL1
      3 942210 PL2
      2 942370 PL2
      2 942190 PL1
      2 942140 PL1
      1 942360 PL1
      1 942230 PL1
      1 933210 PL1

-= v4.0.0-rc2 =-
Freq.   ID #   Paranoia Level
   1490 942110 PL2
    822 932240 PL2
    274 932236 PL2
    130 920240 PL1
     78 942131 PL2
     54 942200 PL2
     25 942440 PL2
     22 942390 PL2
     19 932200 PL2
     17 932380 PL1
     13 942430 PL2
     13 931130 PL2
      8 942410 PL2
      6 933151 PL2
      5 942380 PL2
      5 942150 PL2
      4 942180 PL2
      4 942100 PL1
      4 932235 PL1
      3 942210 PL2
      3 932370 PL1
      2 942520 PL2
      2 942370 PL2
      2 942190 PL1
      2 942151 PL1
      2 942140 PL1
      1 953101 PL2
      1 942362 PL2
      1 942360 PL1
      1 942260 PL2
      1 942130 PL2
```

**Comment by dune73** [comment]
Thanks. These numbers put things in perspective and CRSv4 looks less crazy now - but still room for improvement.

Here is what ChatGPT (+ Link Reader plugin) makes out of the screenshot above (unfortunately it omitted the Delta column):

| Hits | % of reqs. | 1 in X reqs. | Rule ID | Paranoia Level |
|------|------------|-----------|---------|----------------|
| 1490 | 14.90%     | 7         | 942110  | PL2            |
| 822  | 8.22%      | 12        | 932240  | PL2            |
| 274  | 2.74%      | 36        | 932236  | PL2            |
| 130  | 1.30%      | 7         | 920240  | PL1            |
| 78   | 0.78%      | 128       | 942131  | PL2            |
| 54   | 0.54%      | 185       | 942200  | PL2            |
| 35   | 0.25%      | 400       | 942440  | PL2            |
| 22   | 0.22%      | 455       | 942390  | PL2            |
| 19   | 0.19%      | 526       | 932200  | PL2            |
| 17   | 0.17%      | 588       | 932380  | PL1            |
| 13   | 0.13%      | 769       | 942430  | PL2            |
| 13   | 0.13%      | 769       | 931130  | PL2            |
| 8    | 0.08%      | 1250      | 942410  | PL2            |
| 6    | 0.06%      | 1667      | 933151  | PL4            |
| 5    | 0.05%      | 2000      | 942380  | PL2            |
| 5    | 0.05%      | 2000      | 942150  | PL2            |
| 4    | 0.04%      | 2500      | 942180  | PL2            |
| 4    | 0.04%      | 2500      | 942100  | PL1            |
| 4    | 0.04%      | 2500      | 932235  | PLL            |
| 3    | 0.03%      | 3333      | 942210  | PL2            |
| 3    | 0.03%      | 3333      | 932370  | PL1            |
| 2    | 0.02%      | 5000      | 942520  | PL2            |
| 2    | 0.02%      | 5000      | 942370  | PL2            |
| 2    | 0.02%      | 5000      | 942190  | PL1            |
| 2    | 0.02%      | 5000      | 942151  | PL1            |
| 2    | 0.02%      | 5000      | 942140  | PL1            |
| 1    | 0.01%      | 10000     | 953101  | PL2            |
| 1    | 0.01%      | 10000     | 942362  | PL2            |
| 1    | 0.01%      | 10000     | 942360  | PL1            |
| 1    | 0.01%      | 10000     | 942260  | PL2            |
| 1    | 0.01%      | 10000     | 942130  | PL2            |

**Comment by dune73** [comment]
#2602 has additional information on 942200 PL2, also with word lists.

**Comment by RedXanadu** [comment]
Updated the script and re-tested v4.0/dev (CRS Apache containers).

Some new false positives are now detected. The relevant alerts were _not_ counted previously because the script was scanning for the 'paranoia-level/X' tag: this tag may be absent in a log entry if the log message is too long (some parts get truncated, which includes tags.)

_Test date and time: 23 Dec, 23:43 UTC_

| Freq. | ID #   | Paranoia Level |
| ----- | ------ | -------------- |
|  1491 | 942110 | PL2            |
|   822 | 932240 | PL2            |
|   275 | 932236 | PL2            |
|   130 | 920240 | PL1            |
|    94 | 942200 | PL2            |
|    78 | 942131 | PL2            |
|    70 | 942430 | PL2            |
|    25 | 942440 | PL2            |
|    22 | 942390 | PL2            |
|    19 | 932200 | PL2            |
|    17 | 932380 | PL1            |
|    13 | 931130 | PL2            |
|     8 | 942410 | PL2            |
|     6 | 942380 | PL2            |
|     6 | 933151 | PL2            |
|     5 | 942180 | PL2            |
|     5 | 942150 | PL2            |
|     4 | 942100 | PL1            |
|     4 | 932235 | PL1            |
|     3 | 942210 | PL2            |
|     3 | 932370 | PL1            |
|     2 | 942520 | PL2            |
|     2 | 942370 | PL2            |
|     2 | 942190 | PL1            |
|     2 | 942151 | PL1            |
|     2 | 942140 | PL1            |
|     1 | 953101 | PL2            |
|     1 | 942362 | PL2            |
|     1 | 942360 | PL1            |
|     1 | 942260 | PL2            |
|     1 | 942130 | PL2            |

**Comment by RedXanadu** [comment]
Fresh numbers:

_Test date and time: 22 Jan, 20:30 UTC (52cffa5)_

| Freq.  | ID #   | Paranoia Level |
| ------ | ------ | -------------- |
|   1491 | 942110 | PL2            |
|    822 | 932240 | PL2            |
|    254 | 932236 | PL2            |
|    130 | 920240 | **PL1**        |
|     94 | 942200 | PL2            |
|     78 | 942131 | PL2            |
|     70 | 942430 | PL2            |
|     25 | 942440 | PL2            |
|     22 | 942390 | PL2            |
|     19 | 932200 | PL2            |
|     17 | 932380 | **PL1**        |
|     13 | 931130 | PL2            |
|      8 | 942410 | PL2            |
|      6 | 942380 | PL2            |
|      6 | 933151 | PL2            |
|      5 | 942180 | PL2            |
|      5 | 942150 | PL2            |
|      4 | 942100 | **PL1**        |
|      4 | 932235 | **PL1**        |
|      3 | 942210 | PL2            |
|      3 | 932370 | **PL1**        |
|      2 | 942520 | PL2            |
|      2 | 942370 | PL2            |
|      2 | 942190 | **PL1**        |
|      2 | 942151 | **PL1**        |
|      2 | 942140 | **PL1**        |
|      1 | 942362 | PL2            |
|      1 | 942360 | **PL1**        |
|      1 | 942260 | PL2            |
|      1 | 942130 | PL2            |

**Comment by RedXanadu** [comment]
Updated numbers:

_Test date and time: 31 Jan, 17:10 UTC (a36a7e4)_

| Freq.  | ID #   | Paranoia Level |
| ------ | ------ | -------------- |
|    268 | 932236 | PL2            |
|    130 | 920240 | PL2            |
|     94 | 942200 | PL2            |
|     78 | 942131 | PL2            |
|     70 | 942430 | PL2            |
|     25 | 942440 | PL2            |
|     22 | 942390 | PL2            |
|     19 | 932200 | PL2            |
|     17 | 932380 | **PL1**        |
|     13 | 931130 | PL2            |
|     11 | 932240 | PL2            |
|      8 | 942410 | PL2            |
|      6 | 942380 | PL2            |
|      6 | 933151 | PL2            |
|      5 | 942180 | PL2            |
|      5 | 942150 | PL2            |
|      4 | 942100 | **PL1**        |
|      4 | 932235 | **PL1**        |
|      3 | 942210 | PL2            |
|      3 | 932370 | **PL1**        |
|      2 | 942520 | PL2            |
|      2 | 942370 | PL2            |
|      2 | 942190 | **PL1**        |
|      2 | 942151 | **PL1**        |
|      2 | 942140 | **PL1**        |
|      1 | 942362 | PL2            |
|      1 | 942360 | **PL1**        |
|      1 | 942260 | PL2            |
|      1 | 942130 | PL2            |

**Comment by RedXanadu** [comment]
Comparing the current situation (fa4db98) to the situation at the start of this issue:

| Hits   | Rule ID | Paranoia Level |
| ------ | ------- | -------------- |
|    130 | 920240  | :small_red_triangle: PL2 (was PL1)           |
|     94 | 942200  | PL2            |
| :arrow_down: 88 (was 275) | 932236 | PL2            |
|     78 | 942131  | PL2            |
|     70 | 942430  | PL2            |
|     25 | 942440  | PL2            |
|     22 | 942390  | PL2            |
|     19 | 932200  | PL2            |
|     13 | 931130  | PL2            |
| :arrow_down: 11 (was 822) | 932240 | PL2            |
|      8 | 942410  | PL2            |
|      6 | 942380  | PL2            |
|      6 | 933151  | PL2            |
|      5 | 942180  | PL2            |
|      5 | 942150  | PL2            |
|      4 | 942100  | **PL1**        |
|      4 | 932235  | **PL1**        |
|      3 | 942210  | PL2            |
|      2 | 942520  | PL2            |
|      2 | 942370  | PL2            |
|      2 | 942190  | **PL1**        |
|      2 | 942151  | **PL1**        |
|      2 | 942140  | **PL1**        |
| :arrow_down: 2 (was 17) | 932380 | **PL1**        |
|      1 | 942362  | PL2            |
|      1 | 942360  | **PL1**        |
|      1 | 942260  | PL2            |
|      1 | 942130  | PL2            |
| :arrow_down: 0 (was 1491) | 942110 | PL2            |
| :arrow_down: 0 (was 3) | 932370 | **PL1**        |

Comparing the top 5 offending rules at the beginning _(left)_ versus now _(right)_:

| Freq.  | ID #   | Paranoia Level |    | Freq.  | ID #   | Paranoia Level |
| ------ | ------ | -------------- | -- | ------ | ------ | -------------- |
|   1491 | 942110 | PL2            |    |    130 | 920240 | PL2            |
|    822 | 932240 | PL2            |    |     94 | 942200 | PL2            |
|    275 | 932236 | PL2            |    |     88 | 932236 | PL2            |
|    130 | 920240 | PL1            |    |     78 | 942131 | PL2            |
|     94 | 942200 | PL2            |    |     70 | 942430 | PL2            |

**Comment by RedXanadu** [comment]
The numbers show significant improvements!

It really feels like we've rounded off some of the sharp edges of v4.0, making it ready to be released into the world and ready for wider adoption.

We've now addressed all of the rules we agreed to improve before the release, so I'm closing this issue as a big success! :tada:

**Comment by dune73** [comment]
Wow, this is really very cool. Congratulations on all the good work. Imagine the work we removed from the shoulders of our users.
